drivers/hmis/app/models/hmis/form/definition.rb

```diff
@@ -57,6 +59,132 @@ def self.validate_json(json)
     recur_check.call(json)
   end
 
+  def hud_assessment?
+    Types::HmisSchema::Enums::AssessmentRole.as_data_collection_stage(role) != 99
+  end
+
+  def intake?
+    Types::HmisSchema::Enums::AssessmentRole.as_data_collection_stage(role) == 1
+  end
+
+  def exit?
+    Types::HmisSchema::Enums::AssessmentRole.as_data_collection_stage(role) == 3
+  end
+
+  def assessment_date_item
+    @assessment_date_item ||= link_id_item_hash.values.find(&:assessment_date)
+  end
+
+  def find_and_validate_assessment_date(hud_values:, entry_date:, exit_date:)
+    errors = HmisErrors::CustomValidationErrors.new
+    date = nil
+    item = assessment_date_item
+    readable_attribute = item.brief_text || item.text
+
+    if item.present? && hud_values.present?
+      date_string = hud_values[item.link_id]
+
+      if date_string.present?
+        date = HmisUtil::Dates.safe_parse_date(date_string: date_string, reasonable_years_distance: 10)
+        errors.add item.field_name, :invalid, readable_attribute: readable_attribute unless date.present?
+      else
+        errors.add item.field_name, :required, readable_attribute: readable_attribute
+      end
+    elsif hud_assessment?
+      errors.add :assessmentDate, :required, readable_attribute: readable_attribute
+    end
+
+    return [nil, errors.errors] if errors.errors.any?
+
+    # Additional validations for HUD assessment dates to be within appropriate entry/exit bounds
+    if date.present? && hud_assessment?
+      # Ensure assessment date is on or after entry date
+      errors.add item.field_name, :out_of_range, readable_attribute: readable_attribute, message: "must be after entry date (#{entry_date.strftime('%m/%d/%Y')})" if entry_date.present? && !intake? && date < entry_date
+      # Ensure assessment date is on or before exit date
+      errors.add item.field_name, :out_of_range, readable_attribute: readable_attribute, message: "must be before exit date (#{exit_date.strftime('%m/%d/%Y')})" if exit_date.present? && !exit? && date > exit_date
+    end
+
+    date = nil if errors.errors.any?
+
+    [date, errors.errors]
+  end
+
+  def validate_form_values(hud_values, _custom_values = nil)
```

Comment:
This is the actual validation function that validates `hud_values` against the Form Definition.
I still need to work out the exact logic of when to warn. 

Currently on FormItem there are 3 validation-related attributes:
```
field :required, Boolean, 'Whether the item must be included in data results', null: true
field :warn_if_empty, Boolean, 'Whether to show a warning if this question is unanswered', null: true
field :bounds, [Forms::ValueBound], 'Bounds applied to the input value', null: true
```

This is not yet sufficient, its just a start. I do think keeping the "required-ness" outside of ValueBounds is logical, because the bounds apply to the value that actually exists.

I would rather add a separate type for validating **required-ness**, that covers all the scenarios:

* Answer is required to be non-nil
* Answer is required to be non-DNC
* Answer is OK to be nil, but warn about it
* Answer is OK to be DNC, but warn about it
* Any of the above should be combinable with an EnableWhen style object, in order to express something like "IF previousStreetEssh is yes, require dateToStreetEssh as a hard stop. IF previousStreetEssh is no or missing, warn about an empty dateToStreetEssh" (as an example)